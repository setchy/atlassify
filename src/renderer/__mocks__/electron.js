// /**
//  * @typedef {import('../../preload/types.ts').AtlassifyAPI} AtlassifyAPI
//  */

// @ts-ignore
window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

// @ts-ignore
window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

// @ts-ignore
window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

// /** @type {AtlassifyAPI} */
// const mockAtlassifyAPI = {
// 	openExternalLink: jest.fn(),
// 	decryptValue: () => Promise.resolve("decrypted"),
// 	encryptValue: () => Promise.resolve("encrypted"),
// 	platform: {
// 		isLinux: () => false,
// 		isMacOS: () => true,
// 		isWindows: () => false,
// 	},
// 	app: {
// 		hide: jest.fn(),
// 		show: jest.fn(),
// 		quit: jest.fn(),
// 		version: () => Promise.resolve("0.0.1"),
// 	},
// 	zoom: {
// 		getLevel: jest.fn(),
// 		setLevel: jest.fn(),
// 	},
// 	tray: {
// 		updateIcon: jest.fn(),
// 		updateTitle: jest.fn(),
// 		useAlternateIdleIcon: jest.fn(),
// 	},
// 	twemojiDirectory: jest.fn(),
// };

// window.atlassify = mockAtlassifyAPI;

// window.atlassify.decryptValue = Promise.resolve("decrypted");
// window.atlassify.twemojiDirectory = Promise.resolve("/tmp/emoji");

window.atlassify.app = {
  version: Promise.resolve('0.0.1'),
};
