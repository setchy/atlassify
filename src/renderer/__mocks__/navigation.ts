const mockNavigate = vi.fn();

vi.mock('react-router-dom', async () => ({
  ...(await vi.importActual('react-router-dom')),
  useNavigate: () => mockNavigate,
}));

export { mockNavigate };
