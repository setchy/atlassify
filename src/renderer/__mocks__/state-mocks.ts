import {
	type Account,
	type AppearanceSettingsState,
	type AtlassifyState,
	type AuthState,
	type EncryptedToken,
	type FilterSettingsState,
	type Link,
	type NotificationSettingsState,
	OpenPreference,
	type SettingsState,
	type SystemSettingsState,
	Theme,
	type Username,
} from "../types";

export const mockAtlassianCloudAccount: Account = {
	id: "123456789",
	username: "user@atlassify.io" as Username,
	token: "token-123-456" as EncryptedToken,
	name: "Atlassify",
	avatar: "https://avatar.atlassify.io" as Link,
};

export const mockAuth: AuthState = {
	accounts: [mockAtlassianCloudAccount],
};

const mockAppearanceSettings: AppearanceSettingsState = {
	theme: Theme.LIGHT,
	zoomPercentage: 100,
	language: "en",
};

const mockNotificationSettings: NotificationSettingsState = {
	markAsReadOnOpen: true,
	delayNotificationState: false,
	fetchOnlyUnreadNotifications: true,
	groupNotificationsByProduct: false,
	groupNotificationsByProductAlphabetically: false,
};

const mockSystemSettings: SystemSettingsState = {
	openLinks: OpenPreference.FOREGROUND,
	keyboardShortcutEnabled: true,
	showNotificationsCountInTray: true,
	showSystemNotifications: true,
	playSoundNewNotifications: true,
	useAlternateIdleIcon: false,
	openAtStartup: true,
};

const mockFilters: FilterSettingsState = {
	filterTimeSensitive: [],
	filterCategories: [],
	filterReadStates: [],
	filterProducts: [],
	filterActors: [],
};

export const mockSettings: SettingsState = {
	...mockAppearanceSettings,
	...mockNotificationSettings,
	...mockSystemSettings,
	...mockFilters,
};

export const mockState: AtlassifyState = {
	auth: mockAuth,
	settings: mockSettings,
};
