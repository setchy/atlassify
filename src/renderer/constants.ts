export const Constants = {
  STORAGE_KEY: 'atlassify-storage',

  LANGUAGE_STORAGE_KEY: 'atlassify-language',

  ALL_READ_EMOJIS: ['🎉', '🎊', '🥳', '👏', '🙌', '😎', '🏖️', '🚀', '✨', '🏆'],

  FETCH_NOTIFICATIONS_INTERVAL_MS: 60 * 1000, // 1 minute

  REFRESH_ACCOUNTS_INTERVAL_MS: 60 * 60 * 1000, // 1 hour

  MAX_NOTIFICATIONS_PER_ACCOUNT: 999,

  BLOCK_ALIGNMENT_LENGTH_THRESHOLD: 55,
};
