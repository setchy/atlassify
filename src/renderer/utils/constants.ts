export const Constants = {
  REPO_SLUG: 'setchy/atlassify',

  STORAGE_KEY: 'atlassify-storage',

  ALL_READ_EMOJIS: ['🎉', '🎊', '🥳', '👏', '🙌', '😎', '🏖️', '🚀', '✨', '🏆'],

  FETCH_NOTIFICATIONS_INTERVAL: 60000,

  REFRESH_ACCOUNTS_INTERVAL: 3600000,

  MAX_NOTIFICATIONS_PER_ACCOUNT: 999,
};
