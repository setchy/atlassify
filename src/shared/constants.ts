export const APPLICATION = {
  ID: 'com.electron.atlassify',

  NAME: 'Atlassify',

  EVENT_PREFIX: 'atlassify:',

  FIRST_RUN_FOLDER: 'atlassify-first-run',

  WEBSITE: 'https://atlassify.io',

  REPO_SLUG: 'setchy/atlassify',

  DEFAULT_KEYBOARD_SHORTCUT: 'CommandOrControl+Shift+G',

  NOTIFICATION_SOUND: 'notification.wav',

  UPDATE_CHECK_INTERVAL_MS: 24 * 60 * 60 * 1000, // 24 hours
};
