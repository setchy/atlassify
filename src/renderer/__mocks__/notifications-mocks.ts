import type {
  AccountNotifications,
  AtlassifyNotification,
  Link,
} from '../types';
import { PRODUCTS } from '../utils/products';
import { mockAtlassianCloudAccount } from './state-mocks';

export const mockAtlassifyNotifications: AtlassifyNotification[] = [
  {
    id: '138661096',
    order: 0,
    notificationGroup: {
      id: '138661096',
      size: 1,
      additionalActors: [],
    },
    message: '#103: chore(deps): update dependency eslint',
    updated_at: '2020-09-13T21:12:16.662Z',
    entity: {
      title: '#103: chore(deps): update dependency eslint',
      iconUrl:
        'https://bbc-object-storage--frontbucket.us-east-1.prod.public.atl-paas.net/80907f89f58b/img/favicon.ico' as Link,
      url: 'https://bitbucket.org/myorg/notifications-test/pull-requests/103?link_source=platform' as Link,
    },
    path: null,
    product: PRODUCTS.bitbucket,
    actor: {
      displayName: 'atlassify-app',
      avatarURL:
        'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
    },
    category: 'direct',
    readState: 'unread',
    account: mockAtlassianCloudAccount,
    url: 'https://bitbucket.org/myorg/notifications-test/pull-requests/103?link_source=platform' as Link,
    type: 'unknown',
  },
  {
    id: '148827438',
    order: 1,
    notificationGroup: {
      id: '138661096',
      size: 2,
      additionalActors: [
        {
          displayName: 'atlassify-app',
          avatarURL:
            'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
        },
      ],
    },
    message: 'atlassify-app edited your page',
    updated_at: '2020-09-13T21:12:16.662Z',
    entity: {
      title: 'Atlassify Home',
      iconUrl:
        'https://home-static.us-east-1.prod.public.atl-paas.net/confluence-page-icon.svg' as Link,
      url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify/pages/3752329340/Atlassify+Home' as Link,
    },
    path: {
      title: 'Atlassify Space',
      iconUrl: null,
      url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify' as Link,
    },
    product: PRODUCTS.confluence,
    actor: {
      displayName: 'atlassify-app',
      avatarURL:
        'https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/123' as Link,
    },
    category: 'direct',
    readState: 'unread',
    account: mockAtlassianCloudAccount,
    url: 'https://some-tenant.atlassian.net/wiki/spaces/Atlassify/pages/3752329340/Atlassify+Home' as Link,
    type: 'Create',
  },
];

export const mockSingleAtlassifyNotification: AtlassifyNotification =
  mockAtlassifyNotifications[0];

export const mockAccountNotifications: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: mockAtlassifyNotifications,
    hasMoreNotifications: false,
    error: null,
  },
];

export const mockAccountNotificationsWithMorePages: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: mockAtlassifyNotifications,
    hasMoreNotifications: true,
    error: null,
  },
];

export const mockSingleAccountNotifications: AccountNotifications[] = [
  {
    account: mockAtlassianCloudAccount,
    notifications: [mockAtlassifyNotifications[0]],
    hasMoreNotifications: false,
    error: null,
  },
];
