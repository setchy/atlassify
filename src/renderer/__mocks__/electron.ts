import { namespacedEvent } from '../../shared/events';

// Mock `window.Notification`
vi.stubGlobal('Notification', function (title: string) {
  this.title = title;
  return {
    onclick: vi.fn(),
  };
});

// Mock `window.Audio`
vi.stubGlobal(
  'Audio',
  class Audio {
    path: string;
    constructor(path: string) {
      this.path = path;
    }
    play() {}
  },
);

// Mock `window.localStorage`
vi.stubGlobal('localStorage', {
  store: {} as Record<string, string>,
  getItem: vi.fn((key: string) => localStorage.store[key] || null),
  setItem: vi.fn((key: string, item: string) => {
    localStorage.store[key] = item;
  }),
  removeItem: vi.fn(),
});

// Mock `window.alert`
vi.stubGlobal('alert', vi.fn());

export const ipcRenderer = {
  send: vi.fn(),
  on: vi.fn(),
  sendSync: vi.fn(),
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  invoke: vi.fn((channel: string, ..._args: any[]) => {
    switch (channel) {
      case 'get-platform':
        return Promise.resolve('darwin');
      case namespacedEvent('version'):
        return Promise.resolve('0.0.1');
      case namespacedEvent('safe-storage-encrypt'):
        return Promise.resolve('encrypted');
      case namespacedEvent('safe-storage-decrypt'):
        return Promise.resolve('decrypted');
      default:
        return Promise.reject(new Error(`Unknown channel: ${channel}`));
    }
  }),
};

export const shell = {
  openExternal: vi.fn(),
};

export const webFrame = {
  setZoomLevel: vi.fn(),
  getZoomLevel: vi.fn(),
};
