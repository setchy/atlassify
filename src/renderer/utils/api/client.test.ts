import axios from 'axios';

import { mockSingleAtlassifyNotification } from '../../__mocks__/notifications-mocks';
import {
  mockAtlassianCloudAccount,
  mockSettings,
} from '../../__mocks__/state-mocks';
import type { Hostname } from '../../types';
import { Constants } from '../constants';
import {
  checkIfCredentialsAreValid,
  getAuthenticatedUser,
  getCloudIDsForHostnames,
  getJiraProjectTypesByKeys,
  getNotificationsByGroupId,
  getNotificationsForUser,
  markNotificationGroupAsRead,
  markNotificationGroupAsUnread,
  markNotificationsAsRead,
  markNotificationsAsUnread,
} from './client';

jest.mock('axios');

describe('renderer/utils/api/client.ts', () => {
  beforeEach(() => {
    // TODO add more explicit mocks
    (axios as jest.MockedFunction<typeof axios>).mockResolvedValue({
      data: {
        data: {},
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('checkIfCredentialsAreValid - should validate credentials', async () => {
    await checkIfCredentialsAreValid(
      mockAtlassianCloudAccount.username,
      mockAtlassianCloudAccount.token,
    );

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('query Me'),
          variables: undefined,
        },
      }),
    );
  });

  it('getAuthenticatedUser - should fetch authenticated user details', async () => {
    await getAuthenticatedUser(mockAtlassianCloudAccount);

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('query Me'),
          variables: undefined,
        },
      }),
    );
  });

  it('listNotificationsForAuthenticatedUser - should list notifications for user', async () => {
    await getNotificationsForUser(mockAtlassianCloudAccount, mockSettings);

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('query MyNotifications'),
          variables: {
            first: Constants.MAX_NOTIFICATIONS_PER_ACCOUNT,
            flat: !mockSettings.groupNotificationsByTitle,
            readState: 'unread',
          },
        },
      }),
    );
  });

  it('markNotificationsAsRead - should mark notifications as read', async () => {
    await markNotificationsAsRead(mockAtlassianCloudAccount, [
      mockSingleAtlassifyNotification.id,
    ]);

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('mutation MarkAsRead'),
          variables: {
            notificationIDs: [mockSingleAtlassifyNotification.id],
          },
        },
      }),
    );
  });

  it('markNotificationsAsUnread - should mark notifications as unread', async () => {
    await markNotificationsAsUnread(mockAtlassianCloudAccount, [
      mockSingleAtlassifyNotification.id,
    ]);

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('mutation MarkAsUnread'),
          variables: {
            notificationIDs: [mockSingleAtlassifyNotification.id],
          },
        },
      }),
    );
  });

  it('markNotificationGroupAsRead - should mark notification group as read', async () => {
    await markNotificationGroupAsRead(
      mockAtlassianCloudAccount,
      mockSingleAtlassifyNotification.notificationGroup.id,
    );

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('mutation MarkGroupAsRead'),
          variables: {
            groupId: mockSingleAtlassifyNotification.notificationGroup.id,
          },
        },
      }),
    );
  });

  it('markNotificationsAsUnread - should mark notification group as unread', async () => {
    await markNotificationGroupAsUnread(
      mockAtlassianCloudAccount,
      mockSingleAtlassifyNotification.notificationGroup.id,
    );

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('mutation MarkGroupAsUnread'),
          variables: {
            groupId: mockSingleAtlassifyNotification.notificationGroup.id,
          },
        },
      }),
    );
  });

  describe('getNotificationsByGroupId', () => {
    it('getNotificationsByGroupId - should fetch unread notifications by group id', async () => {
      const mockGroupSize = 5;

      await getNotificationsByGroupId(
        mockAtlassianCloudAccount,
        { ...mockSettings, fetchOnlyUnreadNotifications: true },
        mockSingleAtlassifyNotification.notificationGroup.id,
        mockGroupSize,
      );

      expect(axios).toHaveBeenCalledWith(
        expect.objectContaining({
          url: 'https://team.atlassian.net/gateway/api/graphql',
          method: 'POST',
          data: {
            query: expect.stringContaining(
              'query RetrieveNotificationsByGroupId',
            ),
            variables: {
              groupId: mockSingleAtlassifyNotification.notificationGroup.id,
              first: mockGroupSize,
              readState: 'unread',
            },
          },
        }),
      );
    });

    it('getNotificationsByGroupId - should fetch all notifications by group id', async () => {
      const mockGroupSize = 5;

      await getNotificationsByGroupId(
        mockAtlassianCloudAccount,
        { ...mockSettings, fetchOnlyUnreadNotifications: false },
        mockSingleAtlassifyNotification.notificationGroup.id,
        mockGroupSize,
      );

      expect(axios).toHaveBeenCalledWith(
        expect.objectContaining({
          url: 'https://team.atlassian.net/gateway/api/graphql',
          method: 'POST',
          data: {
            query: expect.stringContaining(
              'query RetrieveNotificationsByGroupId',
            ),
            variables: {
              groupId: mockSingleAtlassifyNotification.notificationGroup.id,
              first: mockGroupSize,
              readState: null,
            },
          },
        }),
      );
    });
  });

  it('getJiraProjectTypesByKeys - should fetch project types by keys', async () => {
    const mockProjectKeys = ['PROJ-1', 'PROJ-2'];
    const mockCloudID = 'mock-cloud-id';
    await getJiraProjectTypesByKeys(
      mockAtlassianCloudAccount,
      mockCloudID,
      mockProjectKeys,
    );

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('query RetrieveJiraProjectTypes'),
          variables: {
            cloudId: mockCloudID,
            keys: mockProjectKeys,
          },
        },
      }),
    );
  });

  it('getCloudIDsForHostnames - should fetch cloud ID for hostname', async () => {
    const mockHostnames = ['https://example.atlassian.net'] as Hostname[];

    await getCloudIDsForHostnames(mockAtlassianCloudAccount, mockHostnames);

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        url: 'https://team.atlassian.net/gateway/api/graphql',
        method: 'POST',
        data: {
          query: expect.stringContaining('query RetrieveCloudIDsForHostnames'),
          variables: {
            hostNames: mockHostnames,
          },
        },
      }),
    );
  });
});
