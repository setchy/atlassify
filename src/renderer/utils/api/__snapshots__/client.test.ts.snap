// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/utils/api/client.ts checkIfCredentialsAreValid - should validate credentials 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86dG9rZW4tMTIzLTQ1Ng==",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts getAuthenticatedUser - should fetch authenticated user details 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts getNotificationsByGroupId - should fetch notifications by group id 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts listNotificationsForAuthenticatedUser - should list notifications for user 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts markNotificationGroupAsRead - should mark notification group as read 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts markNotificationsAsRead - should mark notifications as read 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts markNotificationsAsUnread - should mark notification group as unread 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`renderer/utils/api/client.ts markNotificationsAsUnread - should mark notifications as unread 1`] = `
{
  "Accept": "*/*",
  "Authorization": "Basic dXNlckBhdGxhc3NpZnkuaW86ZGVjcnlwdGVk",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;
